types/src/ledger_info.rs

```diff
@@ -373,6 +376,190 @@ impl LedgerInfoWithPartialSignatures {
     }
 }
 
+pub enum SignatureWithStatus {
+    Verified(bls12381::Signature),
+    Unverified(bls12381::Signature),
+}
+
+/// This data structure is used to support the optimistic signature verification feature.
+/// Contains the ledger info and the signatures received on the ledger info from different validators.
+/// Some of the signatures could be verified before inserting into this data structure. Some of the signatures
+/// are not verified. Rather than verifying the signatures immediately, we aggregate all the signatures and
+/// verify the aggregated signature at once. If the aggregated signature is invalid, then we verify each individual
+/// unverified signature and remove the invalid signatures.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct LedgerInfoWithUnverifiedSignatures {
+    ledger_info: LedgerInfo,
+    // These signatures are not yet verified. For efficiency, once enough unverified signatures are collected,
+    // they will be aggregated and verified.
+    unverified_signatures: PartialSignatures,
+    verified_signatures: PartialSignatures,
+}
+
+impl Display for LedgerInfoWithUnverifiedSignatures {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(f, "{}", self.ledger_info)
+    }
+}
+
+impl LedgerInfoWithUnverifiedSignatures {
+    pub fn new(ledger_info: LedgerInfo) -> Self {
+        Self {
+            ledger_info,
+            unverified_signatures: PartialSignatures::empty(),
+            verified_signatures: PartialSignatures::empty(),
+        }
+    }
+
+    pub fn commit_info(&self) -> &BlockInfo {
+        self.ledger_info.commit_info()
+    }
+
+    fn add_verified_signature(
+        &mut self,
+        validator: AccountAddress,
+        signature: bls12381::Signature,
+    ) {
+        self.verified_signatures.add_signature(validator, signature);
+        self.unverified_signatures.remove_signature(validator);
+    }
+
+    fn add_unverified_signature(
+        &mut self,
+        validator: AccountAddress,
+        signature: bls12381::Signature,
+    ) {
+        if self.verified_signatures.contains_voter(&validator) {
+            return;
+        }
+        self.unverified_signatures
+            .add_signature(validator, signature);
+    }
+
+    pub fn add_signature(
+        &mut self,
+        validator: AccountAddress,
+        signature_with_status: SignatureWithStatus,
```

Comment:
I see. After your comment, I tried to `verified: OnceCell<bool>` inside the signature and it didn't work because of the special serde macro. So, I could think of two options:
- Have the signature and verification_status as separate variables like now.
- Create another wrapped signature that has both bls12381::Signature and the verification status.

The first option looks cleaner to me as we don't have to change many use cases of the signature, and the round manager and pending votes/pending order votes inserts the verification status without modifying it. What do you prefer?